pkg/utils/kubernetes/reduce/filters.go

```diff
@@ -86,20 +86,49 @@ func filterServiceAccounts(serviceAccounts []corev1.ServiceAccount) []corev1.Ser
 // filterClusterRoles filters out the ClusterRole to be kept and returns
 // all the ClusterRoles to be deleted.
 // The filtered-out ClusterRole is decided as follows:
-// 1. creationTimestamp (older is better)
+//  1. creationTimestamp (newer is better, because newer ClusterRoles can contain new policy rules)
+//  2. using legacy labels (if present): if a ClusterRole does not have the legacy labels, it is considered newer
+//     and will be kept.
 func filterClusterRoles(clusterRoles []rbacv1.ClusterRole) []rbacv1.ClusterRole {
 	if len(clusterRoles) < 2 {
 		return []rbacv1.ClusterRole{}
 	}
 
-	toFilter := 0
+	newestWithManagedByLabels := -1
+	newestLegacy := -1
 	for i, clusterRole := range clusterRoles {
-		if clusterRole.CreationTimestamp.Before(&clusterRoles[toFilter].CreationTimestamp) {
-			toFilter = i
+		labels := clusterRole.GetLabels()
+
+		_, okManagedBy := labels[consts.GatewayOperatorManagedByLabel]
+		_, okManagedByNs := labels[consts.GatewayOperatorManagedByNamespaceLabel]
+		_, okManagedByName := labels[consts.GatewayOperatorManagedByNameLabel]
+		if okManagedBy && okManagedByNs && okManagedByName {
+			if newestWithManagedByLabels == -1 {
+				newestWithManagedByLabels = i
+				continue
+			}
+
+			if clusterRole.CreationTimestamp.After(clusterRoles[newestWithManagedByLabels].CreationTimestamp.Time) {
+				newestWithManagedByLabels = i
+			}
+			continue
+		}
+
+		if newestLegacy == -1 {
+			newestLegacy = i
+			continue
 		}
+
+		if clusterRole.CreationTimestamp.After(clusterRoles[newestLegacy].CreationTimestamp.Time) {
+			newestLegacy = i
+		}
+		continue
 	}
 
-	return append(clusterRoles[:toFilter], clusterRoles[toFilter+1:]...)
+	if newestWithManagedByLabels != -1 {
+		return append(clusterRoles[:newestWithManagedByLabels], clusterRoles[newestWithManagedByLabels+1:]...)
+	}
+	return append(clusterRoles[:newestLegacy], clusterRoles[newestLegacy+1:]...)
```

Comment:
Here we're exposed to issue described in 
- https://github.com/Kong/gateway-operator/issues/146

Reply:
ACK, let's keep the scope of this PR as small as possible. 146 deserves its own issue. WDYT?